python/rikai/video.py

```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
+import cv2
+import youtube_dl
+
+from rikai.spark.types import (
+    YouTubeVideoType,
+    VideoStreamType,
+)
+from rikai.vision import Image
+
+
+class YouTubeVideo:
+
+    __UDT__ = YouTubeVideoType()
+
+    def __init__(self, vid: str):
+        self.vid = vid
+        self.uri = "https://www.youtube.com/watch?v={0}".format(self.vid)
+        self.embed_url = "http://www.youtube.com/embed/{0}".format(self.vid)
+
+    def __repr__(self) -> str:
+        return "YouTubeVideo({0})".format(self.vid)
+
+    def _repr_html_(self):
+        from IPython.lib.display import YouTubeVideo
+
+        return YouTubeVideo(self.vid)._repr_html_()
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, YouTubeVideo) and self.vid == other.vid
+
+    def get_stream(self, **kwargs) -> "VideoStream":
+        # TODO allow configurable cache dir
+        dl_path = "/tmp/{}.mp4".format(self.vid)
+        opts = {"format": "worst[ext=mp4]", "outtmpl": dl_path}
+        opts.update(kwargs)
+        with youtube_dl.YoutubeDL(opts) as dl:
+            dl.download([self.uri])
+        return VideoStream(dl_path)
+        # TODO copy to S3 instead of local
+
+
+class VideoStream:
+    """VideoStream resource"""
+
+    __UDT__ = VideoStreamType()
+
+    def __init__(self, uri: str):
+        self.uri = uri
+
+    def __repr__(self) -> str:
+        return f"VideoStream(uri={self.uri})"
+
+    def _repr_html_(self):
+        """TODO: codec"""
+        from IPython.display import Video
+
+        return Video(self.uri, embed=True, width=480, height=320)._repr_html_()
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, VideoStream) and self.uri == other.uri
+
+
+class VideoSampler(ABC):
+    """
+    Subclasses will implement different ways to retrieve samples from a given
+    VideoStream.
+    """
+
+    def __init__(self, stream: VideoStream):
+        self.stream = stream
+
+    @abstractmethod
+    def __iter__(self):
+        pass
+
+
+class SingleFrameSampler(VideoSampler):
+    """
+    A simple sampler that just returns one out of every `sample_rate` frames
+    """
+
+    def __init__(
+        self, stream: VideoStream, sample_rate: int, start_frame: int, max_samples: int
+    ):
+        """
+        :param stream: source video to sample from
```

Comment:
We should have a style guide :) 

i used [numpy docstring](https://numpydoc.readthedocs.io/en/latest/format.html) for the rest of the python codebase.  Lets get some consent on docstring format ?

Reply:
These were auto-generated by IntelliJ. I prefer numpy docstring actually, I'll do a docs/comments cleanup pass

Reply:
You can change intellij to set default docstring to numpy style.

Reply:
yeah, i had project settings set before but I created a new one for the new repo and lost it